### DISCUSSION TITLE
Monorepo下api里能否引用其他模块

### DISCUSSION THREAD

**Discussion by xmsz** [web-infra-dev/modern.js#3934] [Q&A] [answered]
背景
- 创建了一个内部模块
- 在api/xxx.ts引用

报错
```
export default function () {
^^^^^^

SyntaxError: Unexpected token 'export'

```
- 如果在src/page.tsx引用这个内部模块就没问题

然后场景是
- 有些包是后端服务的包希望也能共享

**Comment by chenjiahan** [comment]
前后端共用的代码，推荐放到 shared 目录下面：https://modernjs.dev/apis/app/hooks/shared.html

**Reply by xmsz** [reply]
我是Monorepo项目 这个shared目录应该放在哪（根目录还是app项目下）



就是我有个内部包是多个应用后端共享的，不用前后端共享

**Reply by chenjiahan** [reply]
一般放在 app 的目录下面

**Reply by xmsz** [reply]
> 一般放在 app 的目录下面

那我其他应用想要引用这个模块怎么处理

**Reply by chenjiahan** [reply]
那就作为 monorepo 里的一个 package 来引用，用 pnpm 的 workspace 协议来依赖这个 package

**Reply by xmsz** [reply]
> 那就作为 monorepo 里的一个 package 来引用，用 pnpm 的 workspace 协议来依赖这个 package

对 我就是先创建了一个内部模块 然后在api/xx.ts里引用会报错 在src/底下就正常引用

**Reply by yimingjfe** [reply, marked as answer]
目前 BFF 还不支持  monorepo 下的源码开发，需要构建出对应的 cjs 产物，把被引用包的 package.json 中的 `exports` 中 `require` 字段指向这份 cjs 产物（可参考：[Conditional exports](https://nodejs.org/api/packages.html#conditional-exports)）。

这里可以使用 Modern.js 的[模块工程](https://modernjs.dev/module-tools/en/)进行构建，Modern.js 的模块工程可以同时构建出 esm 和 cjs 的产物。（可参考 Modern.js 仓库下的包）
